Dynamic provisioner for AWS EBS

pmorie opened the pull request:
Replaces #15603. 

@thockin @saad-ali @jsafrane @markturansky

saad-ali commented:
Whoops I was looking at https://github.com/kubernetes/kubernetes/pull/15603. Looking at this now

saad-ali commented on pkg/volume/aws_ebs/aws_util.go:
```diff
@@ -107,3 +108,42 @@ func (util *AWSDiskUtil) DetachDisk(c *awsElasticBlockStoreCleaner) error {
+	}
+
+	// convert to MB with rounding
+	volSize := int((c.options.Capacity.Value() + mega - 1) / mega)
```
Could you add a comment on how this work. Are the units assumed?

saad-ali commented on pkg/volume/aws_ebs/aws_ebs.go:
```diff
@@ -349,3 +390,67 @@ func (c *awsElasticBlockStoreCleaner) TearDownAt(dir string) error {
+	}
+	pv.Spec.PersistentVolumeSource.AWSElasticBlockStore.VolumeID = volumeID
+	pv.Spec.Capacity = api.ResourceList{
+		api.ResourceName(api.ResourceStorage): resource.MustParse(fmt.Sprintf("%d", size)),
```
No units suffix? Are they assumed?

saad-ali commented on pkg/volume/aws_ebs/aws_util.go:
```diff
@@ -107,3 +108,42 @@ func (util *AWSDiskUtil) DetachDisk(c *awsElasticBlockStoreCleaner) error {
 }
+
+func (util *AWSDiskUtil) DeleteVolume(d *awsElasticBlockStoreDeleter) error {
+	cloud := d.plugin.host.GetCloudProvider()
```
nit: this code to fetch and cast the cloud provider gets repeated a lot, factor it out into a util method, if possible.

jsafrane replied:
Good point, fixed

saad-ali commented:
A few small comments, but otherwise looks good.

markturansky commented:
@pmorie @saad-ali I suppose the question now is whether we go with these new ones or revert back to @jsafrane's PRs.  We thought we'd be able to get these merged quickly after 14537 and still hit our deadline, but that has already slipped.  I'm thinking we should revert to the original author who owns the feature, since Jan will also be the one improving these same provisioners for many volume types in the next iteration of provisioning.  He'll have all these in shape by Monday morning (he's in Europe and he'll have these rebased before the rest of us get to work).

Hat tip to Paul for stepping up and trying to help get the ball over the goal line.  It's much appreciated.

jsafrane committed:
Use SSD as default volume type.

General purpose SSD ('gp2') volume type is just slighly more expensive than
Magnetic ('standard' / default in AWS), while the performance gain is pretty
significant.

So far, the volumes were created only during testing, where the extra cost
won't make any difference. In future, we plan to introduce QoS classes, where
users could choose SSD/Magnetic depending on their use cases.

'gp2' is just the default volume type for (hopefuly) short period before these
QoS classes are implemented.

pmorie commented:
@markturansky @saad-ali I say we just keep these open.  Jan is an owner in my repo now so he can update these branches with feedback without involving me.

markturansky commented:
Yeah, Jan's already doing that and closed the originals.

jsafrane commented:
Yes, @pmorie, please squash whatever is needed if we get LGTM when I sleep... See you tomorrow.

pmorie commented:
@jsafrane will do

saad-ali commented:
Cool, I'll take a look at this now.

saad-ali commented:
LGTM

saad-ali commented:
Please squash commits, and I'll add the LGTM label

jsafrane committed:
AWS: Use GiB as units for disk sizes.

From some reason, MiBs were used for public functions and AWS cloud provider
recalculated them to GiB. Let's expose what AWS really supports and don't hide
real allocation units.

jsafrane committed:
Implement Creater and Deleter interfaces for AWS EBS.

Also mark the created EBS volumes with tags, so the admin knows
who/what created the volumes.

jsafrane commented:
* To fix review remarks in the other provisioners, I reworked AWS cloud provider to use GiB instead of MiB. AWS EBS internally works with GiB, just the internal Kubernetes cloud provider API from some reason exports MiB.

* Also, I exported `RoundUpSize()` function. Personally, I think the previous version was better. Now GCE and Cinder provisioner pull requests depend on this one (AWS).

jsafrane commented:
Squashed and ready to merge.

jsafrane commented:
@saad-ali: it's ready to merge

saad-ali commented:
ok to merge

pmorie commented:
@saad-ali thanks a lot for the review

pmorie commented:
@k8s-oncall could the on-call dev consider this one for manual merge?  Two more PRs depend on it that we would like to get in for OpenShift 3.1.1

pmorie commented:
Bueller?

On Tuesday, December 15, 2015, Saad Ali <notifications@github.com> wrote:

> ok to merge on green
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/18601#issuecomment-164869905>
> .
>

saad-ali commented:
@jlowdermilk who's covering for scheduled on-call

jsafrane commented:
@saad-ali, thanks for the review! And, what are we waiting for here exactly? Is some bot broken?

pmorie commented:
@k8s-oncall If we could get a manual merge on this, it would be appreciated.  We have two more PRs to get into master before we can put them responsibly into OpenShift for our 3.1.1 release, and we are technically beyond code cutoff.

pmorie commented:
@jsafrane the merge queue is blocked because some E2E groups are broken in master.

pmorie commented:
@jsafrane at least, the queue was blocked yesterday, looks like it's green again now, so hopefully the bot will pick this change up soon.

saad-ali commented:
The build was unstable/broken much of yesterday, so the merge bot was blocked most of the day. Looks like this PR is in 9th in the queue at the moment.

I'll manually merge to unblock you guys. CC @jlowdermilk who's covering for scheduled on-call.

pmorie commented:
@saad-ali thanks a lot, appreciated

jlowdermilk commented:
Sorry I missed your earlier ping(s), was unblocking mergebot. Thanks for merging @saad-ali
